mvcc/backend/backend.go

```diff
@@ -358,7 +358,16 @@ func (b *backend) defrag() error {
 
 	b.batchTx.tx = nil
 
-	tmpdb, err := bolt.Open(b.db.Path()+".tmp", 0600, boltOpenOptions)
+	tmpdbPath := b.db.Path() + ".tmp"
+	if _, err := os.Stat(tmpdbPath); err == nil {
+		// Ensure we start with a clean slate since
+		// db.tmp files can be orphaned if etcd is terminated
+		// before defrag completes
+		if rmErr := os.Remove(tmpdbPath); rmErr != nil && !os.IsNotExist(rmErr) {
+			return fmt.Errorf("failed to remove db.tmp before starting defragmentation: %v", err)
+		}
+	}
+	tmpdb, err := bolt.Open(tmpdbPath, 0600, boltOpenOptions)
```

Comment:
It'd be better if bolt could be persuaded to Create() instead of Open(), then there's no window for something weird to happen on the file system at all.

Reply:
Done. `ioutil.TempFile` does the right thing here (`os.OpenFile(name, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0600)`).